### PR TITLE
Remove Stripe subs. on WP account deletion

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds Stripe and newsletter subscriptions cancellation on account deletion.

### How to test the changes in this Pull Request:

1. Set up a site with Stripe as the Reader Revenue platform and RAS enabled
3. Donate and subscribe to a newsletter as a new reader  
4. Verify in Stripe and ESP that the reader has a Stripe subscription and is subscribed to the newsletter
5. Complete the account deletion flow, starting on My Account page
6. After the account is deleted, verify that the Stripe subscription has been cancelled and the reader unsubscribed from the newsletters

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/reader-revenue/class-stripe-connection.php [hunk 1]
```diff
@@ -1267,6 +1303,24 @@ public static function create_wc_transaction_payload( $customer, $payment ) {
 			'subscribed'         => self::has_customer_opted_in_to_newsletters( $customer ),
 		];
 	}
+
+	/**
+	 * If a WP user is deleted, cancel their Stripe subscriptions.
+	 *
+	 * @param int $user_id User ID.
+	 */
+	public static function cancel_deleted_user_subscriptions( $user_id ) {
```
Comment (miguelpeixe) [inline includes/reader-revenue/class-stripe-connection.php:1312]: The method itself does not take "user deletion" into account so it shouldn't state that.

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
I made a new recurring donation with my email to the Stripe test account, deleted the account, and noticed in the logs the canceling of a subscription (`sub_1LhgC9G3IJS4cjk80RnxUQKl`) that was not the one that had I just created (`sub_1LkXzsG3IJS4cjk8wreYDB4q`)

Turns out there are 2 stripe clients with the same email, which shouldn't happen and is a concern in itself that we should monitor why that would happen. My concern regarding this PR is that the donation was attached to a client and the subscription cancelation happened on another client.

File: includes/reader-revenue/class-stripe-connection.php (hunk 1 above)
Comment (adekbadek) [reply includes/reader-revenue/class-stripe-connection.php]: Fixed in f3c3c6cbd72fed407e81ad6d567e5c0acb3862e9

**Comment by adekbadek** [comment]
> I made a new recurring donation with my email to the Stripe test account, deleted the account, […] the donation was attached to a client and the subscription cancelation happened on another client.

Hm, that should not have happened. The stripe customer ID is tied to a WP used ID either on RAS' [reader verification](https://github.com/Automattic/newspack-plugin/blob/cc26a6d4f41d5cacc777f99d4cafb1ea633a71ca/includes/reader-revenue/class-stripe-connection.php#L1239) or by [matching from user's WC orders](https://github.com/Automattic/newspack-plugin/blob/e1533b6a6d57deba98e986628fc86c991897f742/includes/reader-revenue/my-account/class-woocommerce-my-account.php#L366). I'm guessing the former happened in this case. Can you try to list reproduction steps for this issue to occur? How was the client with the `sub_1LhgC9G3IJS4cjk80RnxUQKl` subscription created?

> Canceling the subscription also triggers an action that recreates the deleted ESP contact with the "Ex-Monthly Donor" meta field, which shouldn't happen.

Should be fixed by b7da828ade68f04b79569b1c3d5648b647916742

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
> Can you try to list reproduction steps for this issue to occur? How was the client with the sub_1LhgC9G3IJS4cjk80RnxUQKl subscription created?

I wasn't aware I had 2 clients with the same email until I tested this PR, so I don't know the steps to reproduce this scenario.

What could be done to further test it would be manually registering 2 clients with the same email under Stripe and see how the subscription creation/cancelation is handled.

This is no longer blocking since I cleaned up the Stripe accounts and was not able to reproduce with a fresh subscription. I'm considering this to not work properly in any circumstance since email is used as the identifier throughout the codebase.

Before approving, I wonder if data protection laws compliance requires us to also delete the customer record. WDYT?

**Comment by yogeshbeniwal** [comment]
> I wonder if data protection laws compliance requires us to also delete the customer record

Keeping financial transaction detail should be ok.

**Comment by adekbadek** [comment]
(some internal discussion in p1663921713036149-slack-C0390J7K0KY)

File: includes/reader-revenue/class-stripe-connection.php [hunk 2]
```diff
@@ -1267,6 +1311,24 @@ public static function create_wc_transaction_payload( $customer, $payment ) {
 			'subscribed'         => self::has_customer_opted_in_to_newsletters( $customer ),
 		];
 	}
+
+	/**
+	 * If a WP user is deleted, cancel their Stripe subscriptions.
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/reader-revenue/class-stripe-connection.php]: Done in b8a1705dc3166dc10d1f968b799d9eadc2eb52e7

**Comment by adekbadek** [comment]
> Before approving, I wonder if data protection laws compliance requires us to also delete the customer record. WDYT?

Internally, we've agreed that this should not block the subscriptions termination on account deletion. Removing PI data from Stripe can be added later. 

This branch was rebased on `master`, can you please re-review @miguelpeixe?

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
Canceling the subscription is working well 👌 

The latest layout changes to "My Account" made the Delete Account button look weird. I suggest a change to something like this:

| Current | Suggestion |
| --- | --- |
| image | image |

**Comment by adekbadek** [comment]
Good idea, done in 8769b380dca829bc75d496add97b0b285e7eda82

**Review by miguelpeixe — APPROVED** [review]
Thank you for revising!

**Pull request merged**